File: backend/app/open_node/domain/node_management.py
from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from open_node.domain.inventory import AgentCommandRead
from open_node.domain.subscriptions import (
    ManagedNodeCreate,
    ManagedNodeRead,
    SubscriptionAccessResponse,
)


class NodeUpdate(BaseModel):
    model_config = {"extra": "forbid"}

    name: str = Field(min_length=1, max_length=120)
    tag: str | None = Field(default=None, max_length=120)
    tags: list[str] = Field(default_factory=list, max_length=24)
    enabled: bool = Field(strict=True)
    parent_id: UUID | None = None
    target_node_id: UUID | None = None
    client_template: dict[str, Any] = Field(default_factory=dict)
    config: dict[str, Any] = Field(default_factory=dict)
    expected_revision: str = Field(pattern=r"^[a-f0-9]{64}$")
    acknowledge_runtime_restart: Literal[True]

    @field_validator("name")
    @classmethod
    def valid_name(cls, value):
        return ManagedNodeCreate.validate_required_text(value)

    @field_validator("tag")
    @classmethod
    def valid_tag(cls, value):
        return ManagedNodeCreate.validate_optional_text(value)

    @field_validator("tags")
    @classmethod
    def valid_tags(cls, value):
        return ManagedNodeCreate.validate_tags(value)

    @field_validator("client_template", "config")
    @classmethod
    def valid_json(cls, value):
        return ManagedNodeCreate.validate_json_objects(value)


class NodeRemoval(BaseModel):
    model_config = {"extra": "forbid"}

    expected_revision: str = Field(pattern=r"^[a-f0-9]{64}$")
    confirm_name: str
    acknowledge_runtime_restart: Literal[True]
    acknowledge_unmanaged_resources: bool = Field(default=False, strict=True)


class NodeReference(BaseModel):
    id: UUID
    name: str


class NodeRemovalServer(BaseModel):
    server_id: UUID
    server_name: str
    inbound_tags: list[str]
    outbound_tags: list[str]
    retained_inbound_tags: list[str] = Field(default_factory=list)
    retained_outbound_tags: list[str] = Field(default_factory=list)
    phase: Literal["withdrawing", "preview", "apply", "inspect", "completed"] = "withdrawing"
    command_id: UUID | None = None
    error: str | None = None
    impact: dict[str, Any] | None = None


class NodeManagementRead(BaseModel):
    node: ManagedNodeRead
    revision: str
    nodes: list[NodeReference]
    plans: list[NodeReference]
    credential_count: int
    servers: list[NodeRemovalServer]
    blockers: list[str]
    warnings: list[str]
    access: list[SubscriptionAccessResponse]
    license_required: Literal[False] = False


class NodeManagementResult(NodeManagementRead):
    commands: list[AgentCommandRead]


class NodeRemovalRead(BaseModel):
    id: UUID
    node_id: UUID
    name: str
    node_ids: list[UUID]
    status: Literal["pending", "failed", "completed"]
    requested_at: datetime
    completed_at: datetime | None = None
    servers: list[NodeRemovalServer]
    warnings: list[str]
    commands: list[AgentCommandRead] = Field(default_factory=list)
    license_required: Literal[False] = False

File: frontend/src/services/node-management.ts
import { authenticatedFetch } from "./auth";
import { requestError } from "./request-error";
import type { AgentCommand } from "../domain/inventory";
import type { ManagedNode, SubscriptionAccessResponse } from "../domain/subscriptions";

export type NodeOperation = "edit" | "remove";
export interface NodeSettings {
  name: string;
  tag: string | null;
  tags: string[];
  enabled: boolean;
  parent_id: string | null;
  target_node_id: string | null;
  client_template: Record<string, unknown>;
  config: Record<string, unknown>;
}
export interface NodeRemovalServer {
  server_id: string;
  server_name: string;
  inbound_tags: string[];
  outbound_tags: string[];
  retained_inbound_tags: string[];
  retained_outbound_tags: string[];
  phase: "withdrawing" | "preview" | "apply" | "inspect" | "completed";
  command_id: string | null;
  error: string | null;
  impact: Record<string, unknown> | null;
}
export interface NodeManagementRead {
  node: ManagedNode;
  revision: string;
  nodes: { id: string; name: string }[];
  plans: { id: string; name: string }[];
  credential_count: number;
  servers: NodeRemovalServer[];
  blockers: string[];
  warnings: string[];
  access: SubscriptionAccessResponse[];
}
export interface NodeManagementResult extends NodeManagementRead { commands: AgentCommand[] }
export interface NodeRemoval {
  id: string;
  node_id: string;
  name: string;
  node_ids: string[];
  status: "pending" | "failed" | "completed";
  servers: NodeRemovalServer[];
  requested_at: string;
  completed_at: string | null;
  warnings: string[];
  commands: AgentCommand[];
}
export function nodeSettings(node: ManagedNode): NodeSettings {
  return {
    name: node.name, tag: node.tag ?? null, tags: [...node.tags], enabled: node.enabled,
    parent_id: node.parent_id ?? null, target_node_id: node.target_node_id ?? null,
    config: structuredClone(node.config), client_template: structuredClone(node.client_template),
  };
}
export function parseNodeObject(value: string, label: string): Record<string, unknown> {
  let parsed: unknown;
  try { parsed = JSON.parse(value); } catch { throw new Error(label + " 必须是有效的 JSON"); }
  if (!parsed || Array.isArray(parsed) || typeof parsed !== "object") throw new Error(label + " 必须是 JSON 对象");
  return parsed as Record<string, unknown>;
}
const base = import.meta.env.VITE_API_BASE_URL ?? "";
async function request<T>(path: string, init?: RequestInit, fetcher = authenticatedFetch): Promise<T> {
  const response = await fetcher(base + "/api/v1" + path, init);
  if (!response.ok) {
    const data = await response.json().catch(() => null);
    const detail = data?.detail;
    throw requestError(detail, `节点请求失败（${response.status}）`);
  }
  return response.json() as Promise<T>;
}
export function getNodeManagement(id: string, fetcher = authenticatedFetch) {
  return request<NodeManagementRead>(`/nodes/${encodeURIComponent(id)}/settings`, undefined, fetcher);
}
export function saveNode(id: string, settings: NodeSettings, revision: string, fetcher = authenticatedFetch) {
  return request<NodeManagementResult>(`/nodes/${encodeURIComponent(id)}/settings`, {
    method: "PUT", headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ...settings, expected_revision: revision, acknowledge_runtime_restart: true }),
  }, fetcher);
}
export function removeNode(id: string, revision: string, name: string, unmanaged: boolean, fetcher = authenticatedFetch) {
  return request<NodeRemoval>(`/nodes/${encodeURIComponent(id)}/remove`, {
    method: "POST", headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ expected_revision: revision, confirm_name: name, acknowledge_runtime_restart: true, acknowledge_unmanaged_resources: unmanaged }),
  }, fetcher);
}
export function getNodeRemoval(id: string, fetcher = authenticatedFetch) {
  return request<NodeRemoval>(`/node-removals/${encodeURIComponent(id)}`, undefined, fetcher);
}
export function retryNodeRemoval(id: string, fetcher = authenticatedFetch) {
  return request<NodeRemoval>(`/node-removals/${encodeURIComponent(id)}/retry`, { method: "POST" }, fetcher);
}
